Kernel/Arch/i386/CPU.cpp

```diff
@@ -444,16 +416,69 @@ static void unimp_trap()
     hang();
 }
 
-void register_irq_handler(u8 irq, IRQHandler& handler)
+GenericInterruptHandler& get_interrupt_handler(u8 interrupt_number)
 {
-    ASSERT(!s_irq_handler[irq]);
-    s_irq_handler[irq] = &handler;
+    ASSERT(s_interrupt_handler[interrupt_number] != nullptr);
+    return *s_interrupt_handler[interrupt_number];
 }
 
-void unregister_irq_handler(u8 irq, IRQHandler& handler)
+static void revert_to_unused_handler(u8 interrupt_number)
 {
-    ASSERT(s_irq_handler[irq] == &handler);
-    s_irq_handler[irq] = nullptr;
+    new UnhandledInterruptHandler(interrupt_number);
+}
+
+void register_generic_interrupt_handler(u8 interrupt_number, GenericInterruptHandler& handler)
+{
+    if (s_interrupt_handler[interrupt_number] != nullptr) {
+        if (s_interrupt_handler[interrupt_number]->purpose() == "Unhandled Interrupt Handler") {
+            s_interrupt_handler[interrupt_number] = &handler;
+            return;
+        } else if (s_interrupt_handler[interrupt_number]->is_shared_handler() && !s_interrupt_handler[interrupt_number]->is_sharing_with_others()) {
+            if (s_interrupt_handler[interrupt_number]->purpose() == "Shared IRQ Handler") {
+                reinterpret_cast<SharedIRQHandler*>(s_interrupt_handler[interrupt_number])->register_handler(handler);
```

Comment:
Since `s_interrupt_handler` is an array of `GenericInterruptHandler*`, we can use `static_cast` here rather than `reinterpret_cast`